### PR TITLE
2 implement data collator

### LINKED ISSUE #2: Implement Data Collator

### PR DESCRIPTION
**Pull request by fouk21**
@ChristosSgouropoulos @vasilistheiou 

Hey once again,

I’ve implemented the data collator class following the existing Whisper collator as closely as possible. In line with the rest of the codebase, it includes docstrings and inline comments explaining the logic.

**Note:** This may need some tweaking. I’m not fully certain that the current padding strategy is optimal or perfectly aligned with Whisper’s training and decoding behavior. However, it is functional for now.

The relevant tensor shapes are documented in the corresponding PR comments.

The collator takes the raw data as input and creates batches with a structure consistent with a dataset containing:
- `input_features`
- `labels`
- `dec_input_ids`

Thanks!

### CODE REVIEW

File: datasets/dataset.py [hunk 1]
```diff
@@ -86,10 +89,78 @@ def __getitem__(self, idx: int) -> dict:
...
+            
+        Returns:
+            Dict[str, torch.Tensor]: Batched and padded tensors:
+                - input_features: Tensor of shape [B, n_mels, time]
+                - dec_input_ids: Tensor of shape [B, max_seq_len]
+                - labels: Tensor of shape [B, max_seq_len]
+        """
+        # Extract audio features (assumed already padded to same length)
+        input_features = [d["input_features"] for d in input_data]
+
```
Comment (fouk21) [inline datasets/dataset.py:139]: len(input_features) == len(dec_input_ids) == len(labels) == batch_size

File: datasets/dataset.py [hunk 2]
```diff
@@ -86,10 +89,78 @@ def __getitem__(self, idx: int) -> dict:
...
+            max(l.size(0) for l in labels),
+            max(d.size(0) for d in dec_input_ids),
+        )
+
+        # Pad labels with -100 so they are ignored by the loss function
+        labels = torch.stack([
+            F.pad(l, (0, max_len - l.size(0)), value=-100)
+            for l in labels
+        ])
+
```
Comment (fouk21) [inline datasets/dataset.py:160]: shape: [batch_size, max_sequence_len]

File: datasets/dataset.py [hunk 3]
```diff
@@ -86,10 +89,78 @@ def __getitem__(self, idx: int) -> dict:
...
+        # Pad decoder input IDs with the Whisper EOT token (50257)
+        dec_input_ids = torch.stack([
+            F.pad(d, (0, max_len - d.size(0)), value=50257)
+            for d in dec_input_ids
+        ])
+
```
Comment (fouk21) [inline datasets/dataset.py:166]: shape: [batch_size, max_sequence_len]

File: datasets/dataset.py [hunk 4]
```diff
@@ -86,10 +89,78 @@ def __getitem__(self, idx: int) -> dict:
...
+        
+        # Convert token lists to tensors
+        labels = [torch.tensor(d["labels"], dtype=torch.long) for d in input_data]
+        dec_input_ids = [
+            torch.tensor(d["dec_input_ids"], dtype=torch.long) for d in input_data
+        ]
+
+        # Stack audio features into a batch tensor
+        input_features = torch.stack(input_features, dim=0)
+
```
Comment (fouk21) [inline datasets/dataset.py:148]: shape: [batch_size, n_mels, time_bins]

**Comment by ChristosSgouropoulos** [comment]
Okay i have also pushed some minor fixes. I am merging
